### PR TITLE
Adding FHMRUVV N3LO splitting functions

### PR DESCRIPTION
**Pull request by giacomomagni**
This PR is imported from: https://github.com/NNPDF/eko_fhmv_private/pull/1

### CODE REVIEW

File: src/ekore/anomalous_dimensions/unpolarized/space_like/as4/__init__.py [hunk 1]
```diff
@@ -6,6 +6,7 @@
 import numba as nb
 import numpy as np
 
+from . import fhmruvv_approximations
```
Comment (felixhekhorn) [inline src/ekore/anomalous_dimensions/unpolarized/space_like/as4/__init__.py:9]:
```suggestion
from . import fhmruvv
```
I guess the acronym is unique enough

File: src/ekore/anomalous_dimensions/unpolarized/space_like/as4/fhmruvv_approximations/__init__.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
+"""The |N3LO| Altarelli-Parisi splitting kernels, from Falcioni, Herzog, Moch, Ruijl, Ueda, Vermaseren, Vogt approximations."""
```
Comment (felixhekhorn) [inline src/ekore/anomalous_dimensions/unpolarized/space_like/as4/fhmruvv_approximations/__init__.py:1]:
```suggestion
"""The |FHMRUVV| |N3LO| Altarelli-Parisi splitting kernels approximations."""
```

File: src/ekore/anomalous_dimensions/unpolarized/space_like/as4/fhmruvv_approximations/__init__.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    Parameters
+    ----------
+    N : complex
+      Mellin moment
+    nf : int
+      Number of active flavors
+    cache: numpy.ndarray
+        Harmonic sum cache
+    variation : tuple
+        |N3LO| anomalous dimension variation ``(gg_var, gq_var, qg_var, qq_var)``
```
Comment (felixhekhorn) [inline src/ekore/anomalous_dimensions/unpolarized/space_like/as4/fhmruvv_approximations/__init__.py:33]:
```suggestion
        |N3LO| anomalous dimension variation ``(gg, gq, qg, qq, nsp, nsm, nsv)``
```

File: src/ekore/anomalous_dimensions/unpolarized/space_like/as4/fhmruvv_approximations/ggg.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
+r"""The unpolarized, space-like anomalous dimension :math:`\gamma_{qg}^{(3)}`."""
```
Comment (felixhekhorn) [inline src/ekore/anomalous_dimensions/unpolarized/space_like/as4/fhmruvv_approximations/ggg.py:1]:
```suggestion
r"""The unpolarized, space-like anomalous dimension :math:`\gamma_{gg}^{(3)}`."""
```

File: src/ekore/anomalous_dimensions/unpolarized/space_like/as4/fhmruvv_approximations/ggq.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+            + 3.4 * bfkl1 * (-(1 / (-1 + n) ** 2))
+            - 158805.0 * 1 / ((-1 + n) * n)
+            + 35098.0 * 1 / n
+            + 87258.0 * (-(1 / n**2))
+            - 4834.1 * lm12(n, S1, S2)
+            + 176.6 * lm13(n, S1, S2, S3)
+        )
+        P3gqApp2 = (
+            P3GQ01
+            # + 5.0*bfkl1* (-(1 / (-1 + n) ** 2))
```
Comment (felixhekhorn) [inline src/ekore/anomalous_dimensions/unpolarized/space_like/as4/fhmruvv_approximations/ggq.py:87]:
```suggestion
            # this was a bug we detected
            # + 5.0*bfkl1* (-(1 / (-1 + n) ** 2))
```
right?

File: src/ekore/anomalous_dimensions/unpolarized/space_like/as4/fhmruvv_approximations/ggg.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+            P3gg01
+            + 5.4 * bfkl1 * -(1 / (-1 + n) ** 2)
+            - 1274800.0 * 1 / ((-1 + n) * n)
+            - 637406.0 * (1 / n - 1 / (1 + n) + 1 / (2 + n) - 1 / (3 + n))
+            - 1314010.0 * (-(1 / n**2))
+            + 177882.0 * lm12m1(n, S1, S2)
+            + 34362.0 * lm13m1(n, S1, S2, S3)
+        )
+    else:
+        raise NotImplementedError("nf^6 is not available at N3LO")
```
Comment (felixhekhorn) [inline src/ekore/anomalous_dimensions/unpolarized/space_like/as4/fhmruvv_approximations/ggg.py:119]:
```suggestion
        raise NotImplementedError("nf=6 is not available at N3LO")
```
likewise for the others ($n_f^k$ actually appears somewhere else)

File: src/ekore/anomalous_dimensions/unpolarized/space_like/as4/fhmruvv_approximations/gnsm.py [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+from ......harmonics.log_functions import lm11, lm11m1, lm12m1, lm13m1
+
+
+@nb.njit(cache=True)
+def gamma_nsm(n, nf, cache, variation):
+    r"""Compute the |N3LO| valence-like non-singlet anomalous dimension.
+
+    The routine is taken from :cite:`Moch:2017uml`.
+
+    The nf^{0,1} leading large-nc contributions and the nf^2 part are
```
Comment (felixhekhorn) [inline src/ekore/anomalous_dimensions/unpolarized/space_like/as4/fhmruvv_approximations/gnsm.py:16]: I'm perfectly happy with keeping the authors original comments, but can we make them verbatim somehow? (else they will render weirdly)

File: src/ekore/anomalous_dimensions/unpolarized/space_like/as4/fhmruvv_approximations/__init__.py (hunk 2 above)
Comment (felixhekhorn) [reply src/ekore/anomalous_dimensions/unpolarized/space_like/as4/fhmruvv_approximations/__init__.py:1]: Maybe also add to the docs here a sentence like "Remember that the authors follow Pegasus convention and so there is an additional global minus sign with respect to our conventions"

**Review by felixhekhorn — CHANGES_REQUESTED** [review]
mostly minor stuff, but
- please check explicitly that you didn't break the LHA benchmarks since they were broken before your last commit and I'm not sure that fixes it (as it was numba complaining)
- we should remember to point them (=FHMRUVV) explicitly to this PR and tell them that they can (and should) start playing on their own

File: src/ekore/anomalous_dimensions/unpolarized/space_like/as4/fhmruvv_approximations/__init__.py (hunk 2 above)
Comment (felixhekhorn) [reply src/ekore/anomalous_dimensions/unpolarized/space_like/as4/fhmruvv_approximations/__init__.py:1]: yes and no - I think it is the usual Pegasus sign: it uses the bare Mellin transform of splitting functions, but we use anomalous dimensions defined as (- Mellin transform of splitting functions) [as the various MVV papers do], no?

File: src/ekore/anomalous_dimensions/unpolarized/space_like/as4/fhmruvv_approximations/ggq.py (hunk 5 above)
Comment (giacomomagni) [reply src/ekore/anomalous_dimensions/unpolarized/space_like/as4/fhmruvv_approximations/ggq.py]: I don't think so... In any case the comment is now gone in the online file:
https://arxiv.org/src/2310.05744v1/anc/xPgq3a.f

File: src/ekore/anomalous_dimensions/unpolarized/space_like/as4/fhmruvv_approximations/__init__.py (hunk 2 above)
Comment (giacomomagni) [reply src/ekore/anomalous_dimensions/unpolarized/space_like/as4/fhmruvv_approximations/__init__.py]: okay let's add a sentence.

**Comment by felixhekhorn** [comment]
> Do we want to have to have `use_fhmruvv` also in the `QED` kernels?

is the effort manageable? if it is too much work, I'd say this is surely no priority ... if it is doable, we have to explicitly mention it when writing the mail ...

**Comment by giacomomagni** [comment]
> is the effort manageable? if it is too much work, I'd say this is surely no priority ... if it is doable, we have to explicitly mention it when writing the mail ...

It's just adding few if statements, I can do it now.

**Review by felixhekhorn — APPROVED** [review]
The moment https://github.com/NNPDF/eko/actions/runs/7489901631/job/20387536993 finished and you checked it, we can merge

**Comment by felixhekhorn** [comment]
Numba failed in https://github.com/NNPDF/eko/actions/runs/7489901631 ... but locally running `poe lha -m "nnlo and sv and ffns"` is just fine ...

I should have almost the same python (3.10.13 (me) vs. 3.10.10 (github)), the same numba (0.57.1), and the same numpy (1.24.4); ocard hash is the same (8beae75) (though theory not, but this was already discussed)

any ideas? @AleCandido maybe?

**Comment by giacomomagni** [comment]
> Numba failed in https://github.com/NNPDF/eko/actions/runs/7489901631 ... but locally running `poe lha -m "nnlo and sv and ffns"` is just fine ...
> 
> I should have almost the same python (3.10.13 (me) vs. 3.10.10 (github)), the same numba (0.57.1), and the same numpy (1.24.4); ocard hash is the same (8beae75) (though theory not, but this was already discussed)
> 
> any ideas? @AleCandido maybe?

It might be due to banana!! 
before `n3lo_ad_variations` had only 4 elements, now 7 can be requited!

**Comment by giacomomagni** [comment]
Okay LHA benchmarks are now working again. [12918f9](https://github.com/NNPDF/eko/pull/335/commits/12918f97b6ef609d414ae8e24ea26ed18a31659d)
Although the proper fix would be a new banana version...

**Comment by felixhekhorn** [comment]
> Okay LHA benchmarks are now working again. [12918f9](https://github.com/NNPDF/eko/pull/335/commits/12918f97b6ef609d414ae8e24ea26ed18a31659d) Although the proper fix would be a new banana version...

good catch! let's merge if https://github.com/NNPDF/eko/actions/runs/7500258647/job/20418576973 goes through so we can move on

**Pull request merged**